File: scala/003/Main.scala
/*
 * ProjectEuler 003
 * Scala / Main.scala
 * copyright (c) 2015 Susisu
 *
 * Scala 2.11.4
 * `scalac Main.scala -d bin`
 */

object Main {
    def main(args: Array[String]) {
        println(calc(600851475143L, 2));
    }

    def calc(n: Long, p: Long): Long = {
        if (n <= 1) p
        else if (n % p == 0) calc(n / p, p)
        else calc(n, p + 1)
    }
}

File: ocaml/003/main.ml
(*
    ProjectEuler 002
    OCaml / main.ml
    copyright (c) 2015 Susisu

    OCaml 4.02.1
    `ocamlc main.ml -o bin/main`
*)

let rec calc n p =
    if n = 1 then
        p
    else if n mod p = 0 then
        calc (n / p) p
    else
        calc n (p + 1);;

let main () =
    begin
        print_int (calc 600851475143 2);
        print_newline ()
    end;;

main ();;
